redash/query_runner/big_query.py

```diff
@@ -246,22 +250,106 @@ def _get_query_result(self, jobs, query):
+        return {'name': table_name, 'columns': columns, 'metadata': metadata}
 
     def _get_columns_schema_column(self, column):
         columns = []
+        metadata = []
+
         if column['type'] == 'RECORD':
             for field in column['fields']:
-                columns.append(u"{}.{}".format(column['name'], field['name']))
+                col_name = u"{}.{}".format(column['name'], field['name'])
+                columns.append(col_name)
+                metadata.append({'name': col_name, 'type': field['type']})
         else:
             columns.append(column['name'])
+            metadata.append({'name': column['name'], 'type': column['type']})
+
+        return columns, metadata
+
+    def _columns_and_samples_to_dict(self, schema, samples):
+        samples_dict = {}
+        if not samples:
+            return samples_dict
+
+        # If a sample exists, its shape/length should be analogous to
+        # the schema provided (i.e their lengths should match up)
+        for i, column in enumerate(schema):
+            if column['type'] == 'RECORD':
+                associated_sample = samples[i]
+                if column.get('mode', None) == 'REPEATED':
+                    # Repeated fields have multiple samples of the same format.
+                    # We only need to show the first one as an example.
+                    associated_sample = [] if len(samples[i]) == 0 else samples[i][0]
+
+                for j, field in enumerate(column['fields']):
+                    col_name = u"{}.{}".format(column['name'], field['name'])
+                    samples_dict[col_name] = None
+                    if type(associated_sample) == list and len(associated_sample) > 0:
+                        samples_dict[col_name] = associated_sample[j]
+            else:
+                samples_dict[column['name']] = samples[i]
+
+        return samples_dict
+
+
+    def _flatten_samples(self, samples):
+        samples_list = []
+        for field in samples:
+            value = field['v']
+            if type(value) == dict:
+                samples_list.append(
+                    self._flatten_samples(value.get('f', []))
+                )
+            elif type(value) == list:
+                samples_list.append(
+                    self._flatten_samples(value)
+                )
+            else:
+                samples_list.append(value)
+
+        return samples_list
+
+    def get_table_sample(self, table_name):
+        if not self.configuration.get('loadSchema', False):
+            return {}
+
+        service = self._get_bigquery_service()
+        project_id = self._get_project_id()
+
+        dataset_id, table_id = table_name.split('.')
 
-        return columns
+        try:
+            sample_response = service.tabledata().list(
+                projectId=project_id,
+                datasetId=dataset_id,
+                tableId=table_id,
+                maxResults=1
+            ).execute()
```

Comment:
This needs to use the `fields` parameter for partial responses and pagination to be a functioning query call, like I did in https://github.com/getredash/redash/pull/3673/files#diff-79a49f870dc6fe9bd78c6c81e5d3b200R267.

The `while` loop is the pagination, you must request the `nextPageToken` to get it to work.

Docs for the "partial response": https://developers.google.com/api-client-library/python/guide/performance#partial-response-fields-parameter

API docs for tabledata listing: https://developers.google.com/apis-explorer/#p/bigquery/v2/bigquery.tabledata.list

API docs for table listing: https://developers.google.com/apis-explorer/#p/bigquery/v2/bigquery.tables.list

Reply:
Hm, I've looked into this a bit and I have a few comments on stuff that works and stuff that I think doesn't quite work:

* Adding `fields` to both these API calls seems to work well and makes sense. Thanks for pointing this out!
* I'm actually not using `tables().list()` but `tables().get()` (https://developers.google.com/apis-explorer/#p/bigquery/v2/bigquery.tables.get) which doesn't return or use a `nextPageToken` I think since it's fetching only 1 table it doesn't need to paginate
* In terms of my use of `tabledata().list()`, I've also added a parameter, `maxResults=1` since I'm only requesting 1 example for a given table. For this reason, I didn't paginate the result. We could potentially add the pagination as a precaution but as far as I can tell, it's not necessary.

Reply:
Okay, that makes sense, apologies for misreading the API that is used here, my fault.

About pagination, the only risk I see is if for some reason we'd remove the maxResult parameter in the future and would not see that pagination is missing, since the code is not expressive enough to show that. So either just add the while loop or add a comment maybe? Not sure what makes more sense 😬

Reply:
Agreed, I'll leave a comment on this.